// auto/dist/plugins/released/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var deepmerge_1 = __importDefault(require("deepmerge"));
var TYPE = '%TYPE';
var VERSION = '%VERSION';
var defaultOptions = {
    label: 'released',
    lockIssues: false,
    message: ":rocket: " + TYPE + " was released in " + VERSION + " :rocket:"
};
var closeIssue = /(?:Close|Closes|Closed|Fix|Fixes|Fixed|Resolve|Resolves|Resolved)\s((?:#\d+(?:,\s)?)+)/gi;
var ReleasedLabelPlugin = /** @class */ (function () {
    function ReleasedLabelPlugin(options) {
        if (options === void 0) { options = {}; }
        this.name = 'Released Label';
        this.options = deepmerge_1.default(defaultOptions, options);
    }
    ReleasedLabelPlugin.prototype.apply = function (auto) {
        var _this = this;
        auto.hooks.modifyConfig.tap(this.name, function (config) {
            config.labels.released = config.labels.released || {
                name: 'released',
                description: 'This issue/pull request has been released.'
            };
            return config;
        });
        auto.hooks.afterRelease.tapPromise(this.name, function (newVersion, commits) { return __awaiter(_this, void 0, void 0, function () {
            var head, isSkipped;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!newVersion) {
                            return [2 /*return*/];
                        }
                        if (auto.args.dryRun) {
                            return [2 /*return*/];
                        }
                        head = commits[0];
                        if (!head) {
                            return [2 /*return*/];
                        }
                        isSkipped = head.labels.find(function (label) {
                            return auto.release.options.skipReleaseLabels.includes(label);
                        });
                        if (isSkipped) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, Promise.all(commits.map(function (commit) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, this.addReleased(auto, commit, newVersion)];
                            }); }); }))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    };
    ReleasedLabelPlugin.prototype.addReleased = function (auto, commit, newVersion) {
        return __awaiter(this, void 0, void 0, function () {
            var messages, pr, commitsInPr, issues;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        messages = [commit.subject];
                        if (!commit.pullRequest) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.addCommentAndLabel(auto, newVersion, commit.pullRequest.number)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, auto.git.getPullRequest(commit.pullRequest.number)];
                    case 2:
                        pr = _a.sent();
                        pr.data.body.split('\n').map(function (line) { return messages.push(line); });
                        return [4 /*yield*/, auto.git.getCommitsForPR(commit.pullRequest.number)];
                    case 3:
                        commitsInPr = _a.sent();
                        commitsInPr.map(function (c) { return messages.push(c.commit.message); });
                        _a.label = 4;
                    case 4:
                        issues = messages
                            .map(function (message) { return message.match(closeIssue); })
                            .filter(function (r) { return !!r; })
                            .reduce(function (all, arr) { return __spread(all, arr); }, [])
                            .map(function (issue) { return issue.match(/#(\d+)/i); })
                            .filter(function (r) { return !!r; })
                            .map(function (match) { return Number(match[1]); });
                        return [4 /*yield*/, Promise.all(issues.map(function (issue) { return __awaiter(_this, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.addCommentAndLabel(auto, newVersion, issue, true)];
                                        case 1:
                                            _a.sent();
                                            if (!this.options.lockIssues) return [3 /*break*/, 3];
                                            return [4 /*yield*/, auto.git.lockIssue(issue)];
                                        case 2:
                                            _a.sent();
                                            _a.label = 3;
                                        case 3: return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 5:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ReleasedLabelPlugin.prototype.addCommentAndLabel = function (auto, newVersion, prOrIssue, isIssue) {
        if (isIssue === void 0) { isIssue = false; }
        return __awaiter(this, void 0, void 0, function () {
            var comment, labels;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        comment = this.createReleasedComment(isIssue, newVersion);
                        return [4 /*yield*/, auto.git.createComment(comment, prOrIssue, 'released')];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, auto.git.getLabels(prOrIssue)];
                    case 2:
                        labels = _a.sent();
                        if (!!labels.includes(this.options.label)) return [3 /*break*/, 4];
                        return [4 /*yield*/, auto.git.addLabelToPr(prOrIssue, this.options.label)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ReleasedLabelPlugin.prototype.createReleasedComment = function (isIssue, version) {
        return this.options.message
            .replace(TYPE, isIssue ? 'Issue' : 'PR')
            .replace(VERSION, version);
    };
    return ReleasedLabelPlugin;
}());
exports.default = ReleasedLabelPlugin;
