// auto/dist/utils/make-hooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tapable_1 = require("tapable");
exports.makeHooks = function () { return ({
    beforeRun: new tapable_1.SyncHook(['config']),
    modifyConfig: new tapable_1.SyncWaterfallHook(['config']),
    beforeShipIt: new tapable_1.SyncHook([]),
    afterShipIt: new tapable_1.AsyncParallelHook(['version', 'commits']),
    afterRelease: new tapable_1.AsyncParallelHook(['version', 'commits']),
    onCreateRelease: new tapable_1.SyncHook(['options']),
    onCreateChangelog: new tapable_1.SyncHook(['changelog']),
    onCreateLogParse: new tapable_1.SyncHook(['logParse']),
    getAuthor: new tapable_1.AsyncSeriesBailHook([]),
    getPreviousVersion: new tapable_1.AsyncSeriesBailHook(['prefixRelease']),
    getRepository: new tapable_1.AsyncSeriesBailHook([]),
    version: new tapable_1.AsyncParallelHook(['version']),
    afterVersion: new tapable_1.AsyncParallelHook([]),
    publish: new tapable_1.AsyncParallelHook(['version']),
    afterPublish: new tapable_1.AsyncParallelHook([])
}); };
exports.makeReleaseHooks = function () { return ({
    onCreateChangelog: new tapable_1.SyncHook(['changelog']),
    onCreateLogParse: new tapable_1.SyncHook(['logParse'])
}); };
exports.makeLogParseHooks = function () { return ({
    parseCommit: new tapable_1.AsyncSeriesWaterfallHook(['commit']),
    omitCommit: new tapable_1.AsyncSeriesBailHook(['commit'])
}); };
exports.makeChangelogHooks = function () { return ({
    renderChangelogLine: new tapable_1.AsyncSeriesBailHook(['commits', 'lineRender']),
    renderChangelogTitle: new tapable_1.AsyncSeriesBailHook(['commits', 'lineRender']),
    renderChangelogAuthor: new tapable_1.AsyncSeriesBailHook([
        'author',
        'commit',
        'options'
    ]),
    renderChangelogAuthorLine: new tapable_1.AsyncSeriesBailHook(['author', 'user'])
}); };
