// auto/dist/plugins/chrome/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = __importStar(require("fs"));
var util_1 = require("util");
var semver_1 = require("semver");
var exec_promise_1 = __importDefault(require("../../utils/exec-promise"));
var readFile = util_1.promisify(fs.readFile);
var writeFile = util_1.promisify(fs.writeFile);
function getManifest(path) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _b = (_a = JSON).parse;
                    return [4 /*yield*/, readFile(path, 'utf-8')];
                case 1: return [2 /*return*/, _b.apply(_a, [_c.sent()])];
            }
        });
    });
}
var ChromeWebStorePlugin = /** @class */ (function () {
    function ChromeWebStorePlugin(config) {
        this.name = 'Chrome Web Store';
        this.id = config.id || process.env.EXTENSION_ID;
        this.manifest = config.manifest || 'manifest.json';
        this.build = config.build || process.env.EXTENSION_BUILD || 'extension.zip';
    }
    ChromeWebStorePlugin.prototype.reportWarning = function (auto, message) {
        auto.logger.log.warn(this.name + ": " + message);
    };
    ChromeWebStorePlugin.prototype.apply = function (auto) {
        var _this = this;
        auto.hooks.beforeRun.tap(this.name, function () {
            if (!_this.id) {
                _this.reportWarning(auto, 'EXTENSION_ID environment variable must be set or "id" in autorc');
            }
            if (!fs.existsSync(_this.manifest)) {
                _this.reportWarning(auto, "\"" + _this.manifest + "\" must exist to publish. Or provide a custom path in you autorc");
            }
            if (!fs.existsSync(_this.build)) {
                _this.reportWarning(auto, "Path to either a zip file, or a directory to be zipped at " + _this.build);
            }
            // Secrets
            if (!process.env.CLIENT_ID) {
                _this.reportWarning(auto, 'CLIENT_ID environment variable must be set');
            }
            if (!process.env.CLIENT_SECRET) {
                _this.reportWarning(auto, 'CLIENT_SECRET environment variable must be set');
            }
            if (process.env.REFRESH_TOKEN) {
                return;
            }
            _this.reportWarning(auto, 'REFRESH_TOKEN environment variable must be set');
        });
        auto.hooks.beforeShipIt.tap(this.name, function () {
            if (process.env.CLIENT_ID &&
                process.env.CLIENT_SECRET &&
                process.env.REFRESH_TOKEN &&
                fs.existsSync(_this.manifest) &&
                _this.id &&
                fs.existsSync(_this.build)) {
                return;
            }
            throw new Error("You don't have all necessary config set up! Check the warnings.");
        });
        auto.hooks.getAuthor.tapPromise(this.name, function () { return __awaiter(_this, void 0, void 0, function () {
            var manifest, author;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        auto.logger.verbose.info(this.name + ": Getting author information from package.json");
                        return [4 /*yield*/, getManifest(this.manifest)];
                    case 1:
                        manifest = _a.sent();
                        if (manifest.author) {
                            author = manifest.author;
                            auto.logger.log.info(this.name + ": You will need to manually set the author email in the .autorc.");
                            return [2 /*return*/, author];
                        }
                        auto.logger.log.info(this.name + ": No author found in manifest You will need to manually set the author email and name in the .autorc.");
                        return [2 /*return*/];
                }
            });
        }); });
        auto.hooks.getPreviousVersion.tapPromise(this.name, function (prefixRelease) { return __awaiter(_this, void 0, void 0, function () {
            var manifest, version;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, getManifest(this.manifest)];
                    case 1:
                        manifest = _a.sent();
                        version = prefixRelease(manifest.version);
                        auto.logger.verbose.info(this.name + ": Got previous version from package.json", version);
                        return [2 /*return*/, version];
                }
            });
        }); });
        auto.hooks.getRepository.tap(this.name, function () {
            auto.logger.verbose.info(_this.name + ": Manifest.json has no field to store repo information. Use the .autorc to specify these options.");
        });
        auto.hooks.version.tapPromise(this.name, function (version) { return __awaiter(_this, void 0, void 0, function () {
            var manifest;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, getManifest(this.manifest)];
                    case 1:
                        manifest = _a.sent();
                        manifest.version = semver_1.inc(manifest.version, version);
                        return [4 /*yield*/, writeFile(this.manifest, JSON.stringify(manifest, undefined, 2))];
                    case 2:
                        _a.sent();
                        // commit new version
                        return [4 /*yield*/, exec_promise_1.default('git', ['add', this.manifest])];
                    case 3:
                        // commit new version
                        _a.sent();
                        return [4 /*yield*/, exec_promise_1.default('git', [
                                'commit',
                                '-m',
                                "\"Bump version to " + manifest.version + " [skip ci]\"",
                                '--no-verify'
                            ])];
                    case 4:
                        _a.sent();
                        return [4 /*yield*/, exec_promise_1.default('git', ['tag', manifest.version])];
                    case 5:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        auto.hooks.publish.tapPromise(this.name, function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: 
                    // publish extension
                    return [4 /*yield*/, exec_promise_1.default('webstore', [
                            'upload',
                            '--extension-id',
                            this.id,
                            '--source',
                            this.build,
                            '--auto-publish'
                        ])];
                    case 1:
                        // publish extension
                        _a.sent();
                        // push to github
                        return [4 /*yield*/, exec_promise_1.default('git', [
                                'push',
                                '--follow-tags',
                                '--set-upstream',
                                'origin',
                                '$branch'
                            ])];
                    case 2:
                        // push to github
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    };
    return ChromeWebStorePlugin;
}());
exports.default = ChromeWebStorePlugin;
