// auto/dist/init.js
"use strict";
// tslint:disable no-unnecessary-type-annotation
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var dedent_1 = __importDefault(require("dedent"));
var enquirer_1 = require("enquirer");
var fs = __importStar(require("fs"));
var path = __importStar(require("path"));
var util_1 = require("util");
var release_1 = require("./release");
var writeFile = util_1.promisify(fs.writeFile);
var isObject = function (value) { return typeof value === 'object' && value !== null; };
function getFlags() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, enquirer_1.prompt([
                    {
                        type: 'input',
                        name: 'repo',
                        message: 'GitHub Project Repository (press enter to use package definition)'
                    },
                    {
                        type: 'input',
                        name: 'owner',
                        message: 'GitHub Project Owner (press enter to use package definition)'
                    },
                    {
                        type: 'confirm',
                        name: 'noVersionPrefix',
                        message: 'Use the version as the tag without the `v` prefix?',
                        initial: 'no'
                    },
                    {
                        type: 'input',
                        name: 'jira',
                        message: 'Jira base URL (press enter to skip)'
                    },
                    {
                        type: 'input',
                        name: 'githubApi',
                        message: 'GitHub API to use (press enter to use public)'
                    },
                    {
                        type: 'confirm',
                        name: 'onlyPublishWithReleaseLabel',
                        message: 'Only bump version if `release` label is on pull request',
                        initial: 'no'
                    },
                    {
                        type: 'input',
                        name: 'name',
                        message: 'Git name to commit and release with (press enter to use package definition)'
                    },
                    {
                        type: 'input',
                        name: 'email',
                        message: 'Git email to commit with (press enter to use package definition)'
                    }
                ])];
        });
    });
}
function getCustomLabels(onlyLabels) {
    if (onlyLabels === void 0) { onlyLabels = false; }
    return __awaiter(this, void 0, void 0, function () {
        var _a, _b, _c, useCustomChangelogTitles, _d, customLabels, labels, i, _e, labelName, labelDef, response, _f, name, title, description, newLabel, getAnotherTitle, response, _g, name, title, description;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    if (!onlyLabels) return [3 /*break*/, 1];
                    _d = { value: onlyLabels };
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, enquirer_1.prompt({
                        type: 'confirm',
                        name: 'value',
                        message: 'Would you like to use custom labels?',
                        initial: 'no'
                    })];
                case 2:
                    _d = _h.sent();
                    _h.label = 3;
                case 3:
                    useCustomChangelogTitles = _d;
                    customLabels = {};
                    if (!useCustomChangelogTitles.value) return [3 /*break*/, 6];
                    labels = Object.entries(release_1.defaultLabelDefinition);
                    i = 0;
                    _h.label = 4;
                case 4:
                    if (!labels[i]) return [3 /*break*/, 6];
                    _e = __read(labels[i++], 2), labelName = _e[0], labelDef = _e[1];
                    return [4 /*yield*/, enquirer_1.prompt({
                            type: 'snippet',
                            name: 'value',
                            message: "Customize the " + labelName + " label:",
                            initial: labelDef,
                            // @ts-ignore
                            template: labelName === 'release' || labelName === 'skip-release'
                                ? dedent_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n                label:  #{name}\n                desc:   #{description}\n              "], ["\n                label:  #{name}\n                desc:   #{description}\n              "]))) : dedent_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n                label:  #{name}\n                title:  #{title}\n                desc:   #{description}\n              "], ["\n                label:  #{name}\n                title:  #{title}\n                desc:   #{description}\n              "])))
                        })];
                case 5:
                    response = _h.sent();
                    _f = response.value.values, name = _f.name, title = _f.title, description = _f.description;
                    newLabel = {};
                    if (name !== labelDef.name) {
                        newLabel.name = name;
                    }
                    if (title !== labelDef.title) {
                        newLabel.title = title;
                    }
                    if (description !== labelDef.description) {
                        newLabel.description = description;
                    }
                    if (Object.keys(newLabel).length === 1 && newLabel.name) {
                        customLabels = __assign({}, customLabels, (_a = {}, _a[labelName] = name, _a));
                    }
                    else if (Object.keys(newLabel).length !== 0) {
                        customLabels = __assign({}, customLabels, (_b = {}, _b[labelName] = newLabel, _b));
                    }
                    return [3 /*break*/, 4];
                case 6: return [4 /*yield*/, enquirer_1.prompt({
                        type: 'confirm',
                        name: 'value',
                        message: 'Would you like to add additional labels?',
                        initial: 'no'
                    })];
                case 7:
                    getAnotherTitle = _h.sent();
                    _h.label = 8;
                case 8:
                    if (!getAnotherTitle.value) return [3 /*break*/, 11];
                    return [4 /*yield*/, enquirer_1.prompt({
                            type: 'snippet',
                            name: 'value',
                            message: 'Add another label:',
                            // @ts-ignore
                            template: dedent_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n        label:  #{name}\n        title:  #{title}\n        desc:   #{description}\n      "], ["\n        label:  #{name}\n        title:  #{title}\n        desc:   #{description}\n      "]))),
                            validate: function (state) {
                                if (!state.values.name) {
                                    return 'Label is required for new label';
                                }
                                if (!state.values.title) {
                                    return 'Title is required for new label';
                                }
                                if (!state.values.description) {
                                    return 'Description is required for new label';
                                }
                                return true;
                            }
                        })];
                case 9:
                    response = _h.sent();
                    _g = response.value.values, name = _g.name, title = _g.title, description = _g.description;
                    customLabels = __assign({}, customLabels, (_c = {}, _c[name] = { name: name, title: title, description: description }, _c));
                    return [4 /*yield*/, enquirer_1.prompt({
                            type: 'confirm',
                            name: 'value',
                            message: 'Would you like to add another?',
                            initial: 'no'
                        })];
                case 10:
                    getAnotherTitle = _h.sent();
                    return [3 /*break*/, 8];
                case 11: return [2 /*return*/, customLabels];
            }
        });
    });
}
function init(_a, logger) {
    var onlyLabels = _a.onlyLabels, dryRun = _a.dryRun;
    return __awaiter(this, void 0, void 0, function () {
        var flags, _b, labels, autoRc, jsonString;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!onlyLabels) return [3 /*break*/, 1];
                    _b = {};
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, getFlags()];
                case 2:
                    _b = _c.sent();
                    _c.label = 3;
                case 3:
                    flags = _b;
                    return [4 /*yield*/, getCustomLabels(onlyLabels)];
                case 4:
                    labels = _c.sent();
                    autoRc = Object.entries(__assign({}, flags, { labels: labels })).reduce(function (all, _a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        var _c;
                        if (value === '' ||
                            value === false ||
                            (isObject(value) && Object.keys(value).length === 0)) {
                            return all;
                        }
                        return __assign({}, all, (_c = {}, _c[key] = value, _c));
                    }, {});
                    if (Object.keys(autoRc).length === 0) {
                        return [2 /*return*/];
                    }
                    jsonString = JSON.stringify(autoRc, undefined, 2);
                    if (!dryRun) return [3 /*break*/, 5];
                    logger.log.note("Initialization options would be:\n" + jsonString);
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, writeFile(path.join(process.cwd(), '.autorc'), jsonString)];
                case 6:
                    _c.sent();
                    _c.label = 7;
                case 7: return [2 /*return*/];
            }
        });
    });
}
exports.default = init;
var templateObject_1, templateObject_2, templateObject_3;
