// auto/dist/semver.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var SEMVER;
(function (SEMVER) {
    SEMVER["major"] = "major";
    SEMVER["premajor"] = "premajor";
    SEMVER["minor"] = "minor";
    SEMVER["preminor"] = "preminor";
    SEMVER["patch"] = "patch";
    SEMVER["prepatch"] = "prepatch";
    SEMVER["noVersion"] = "";
})(SEMVER || (SEMVER = {}));
exports.default = SEMVER;
function getHigherSemverTag(left, right) {
    if (left === SEMVER.major || right === SEMVER.major) {
        return SEMVER.major;
    }
    if (left === SEMVER.minor || right === SEMVER.minor) {
        return SEMVER.minor;
    }
    return SEMVER.patch;
}
exports.getHigherSemverTag = getHigherSemverTag;
function calculateSemVerBump(labels, labelMap, _a) {
    var _b = _a === void 0 ? {} : _a, onlyPublishWithReleaseLabel = _b.onlyPublishWithReleaseLabel, _c = _b.skipReleaseLabels, skipReleaseLabels = _c === void 0 ? [] : _c;
    var labelSet = new Set();
    if (!skipReleaseLabels.includes(labelMap.get('skip-release'))) {
        skipReleaseLabels.push(labelMap.get('skip-release'));
    }
    labels.map(function (pr) {
        pr.forEach(function (label) {
            var userLabel = __spread(labelMap.entries()).find(function (pair) { return pair[1] === label; });
            labelSet.add(userLabel ? userLabel[0] : label);
        });
    });
    var skipRelease = false;
    var isPrerelease = false;
    if (labels.length > 0 && labels[0].length > 0) {
        isPrerelease = labels[0].includes(labelMap.get('prerelease'));
        skipRelease = onlyPublishWithReleaseLabel
            ? !labels[0].includes(labelMap.get('release'))
            : !!labels[0].find(function (label) { return skipReleaseLabels.includes(label); });
    }
    var version = __spread(labelSet).reduce(getHigherSemverTag, SEMVER.patch);
    if (skipRelease) {
        return SEMVER.noVersion;
    }
    if (isPrerelease) {
        if (version === SEMVER.major) {
            return SEMVER.premajor;
        }
        if (version === SEMVER.minor) {
            return SEMVER.preminor;
        }
        return SEMVER.prepatch;
    }
    return version;
}
exports.calculateSemVerBump = calculateSemVerBump;
